File: bubbleSort/bubbleSort.py
#Optimized Bubble sort in python
def bubbleSort(array, mLess = lambda x, y : x < y):

    #loop to access each array element
    for i in range(len(array)):
        
        #keep track of swapping
        swapped = False
        
        #loop to compare elements
        for j in range(len(array) -i -1):

            #compare two adjacent elements using mLess comparator
            if mLess(array[j +1], array[j]):

                #swapping elements occurred
                temp = array[j]
                array[j] = array[j +1]
                array[j +1] = temp
                swapped = True
                    
            #check if the array is already sorted
            if not swapped:
                break

File: bubbleSort/bubbleSort.cpp
#include <bits/stdc++.h>
using namespace std;

template <typename T,
           typename Comp = std::less<T> >

//Optimized Bubble sort in C++
void bubbleSort(std::vector<T> &v, Comp mLess = Comp()){

	//loop to access each array element
	for(size_t i = 0; i < v.size() -1; ++i){

		//check if swapping occurs
		bool swapped = false;

		//loop to compare two elements
		for(size_t j = 0; j < v.size() -i -1; ++j){
			//compare two adjacent elements using mLess comparator

			if(mLess(v[j +1], v[j])){

				//swapping element occurs if elements aren't in intended order
				std::swap(v[j], v[j +1]);
				swapped = true;
			}
		}

		//check if an array is sorted
		if(!swapped){
			break;
		}
	}
}